src/PhpParser/NodeVisitor/NodeConnectingVisitor.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\Core\PhpParser\NodeVisitor;
+
+use PhpParser\Node;
+use PhpParser\NodeVisitorAbstract;
+use Rector\NodeTypeResolver\Node\AttributeKey;
+
+/**
+ * Visitor that connects a child node to its parent node
+ * as well as its sibling nodes, with verify previous node is not equal to current node.
+ *
+ * inspired by https://github.com/nikic/PHP-Parser/blob/0ffddce52d816f72d0efc4d9b02e276d3309ef01/lib/PhpParser/NodeVisitor/NodeConnectingVisitor.php
+ */
+final class NodeConnectingVisitor extends NodeVisitorAbstract
+{
+    /**
+     * @var Node[]
+     */
+    private array $stack = [];
+
+    private ?Node $node = null;
+
+    public function beforeTraverse(array $nodes)
+    {
+        $this->stack = [];
+        $this->node = null;
+
+        return null;
+    }
+
+    public function enterNode(Node $node)
+    {
+        if ($this->stack !== []) {
+            $node->setAttribute(AttributeKey::PARENT_NODE, $this->stack[count($this->stack) - 1]);
+        }
+
+        if ($this->node instanceof Node
+            && $this->node !== $node
+            && $this->node->getAttribute(AttributeKey::PARENT_NODE) === $node->getAttribute(
+                AttributeKey::PARENT_NODE
+            )) {
+            $node->setAttribute(AttributeKey::PREVIOUS_NODE, $this->node);
```

Comment:
I think if you add such a visitor which connects each node which its prev/next and parent you will prevent garbage collection.

So you effectively work against the base idea of https://phpstan.org/blog/preprocessing-ast-for-custom-rules


It would be better if would not need at least some of those pointers

Reply:
I am thinking that reduce next-prev usage is possible, but not actually remove it, since there is a dynamic case which we rely on what next node or by prev node, which the position can be indirect or jump.

Rector not only read node, but modify the position.

That can be reduced, but not actually be removed, at least for now.

This custom visitor is only applied after refactor on

https://github.com/rectorphp/rector-src/blob/5c1be93b80c3f7bfd8bd158ba0a55501a7acaefc/src/Rector/AbstractRector.php#L441